### PR TITLE
Added #9398: add checkout / checkin functionality to licenses in API REST

### PR DESCRIPTION
**Pull request by deivishome**
app/Http/Controllers/Api/LicensesController.php --> added functions to checkout (to user and to asset) and checkin based on web requests and adapted to REST API.
app/Http/Requests/LicenseCheckoutRequest.php --> added data needed from the post request to be parsed to the main function in LicenseController.php
app/Models/License.php --> added function to return if there is any available seat for the checkout process in the API REST.
routes/api.php --> Added 2 new routes in route group of Licenses to allow checkout and checkin

Warning: It does not generate any event.
Returns: success / error

# Description

It does not fix an issue, it brings new option to Licenses management from API Rest.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A
- [ ] Test B

**Test Configuration**:
* PHP version: 7.4
* MySQL version: MariaDB 10.5.9
* Webserver version Apahce 2.4.46
* OS version: Ubuntu 18.04.5 x64

# Checklist:

- [x] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [x] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by deivishome** [comment]
Tested in my production snipe-it, from Postman and .NET Framework external application using RESTSharp client.
Now Licenses has the options tobe checkedout and checked in from API Rest.

**Comment by deivishome** [comment]
Attaching captures of postman
scshot-checkinlicense
scsh-assignlicensetoasset
scsh-assignlicensetouser

**Comment by snipe** [comment]
Hi there - thanks for this! Can you please re-target this PR to point to the `develop` branch, per [our Contributing documentation](https://snipe-it.readme.io/docs/contributing-overview)? 

You don't need to close and re-open. After you create a pull request, you can modify the base branch so that the changes in the pull request are compared against a different branch. By changing the base branch of your original pull request rather than opening a new one with the correct base branch, you’ll be able to keep valuable work and discussion.

mhuxq

Thanks!

**Comment by deivishome** [comment]
OK changed branch to develope

**Comment by snipe** [comment]
@deivishome I'm still seeing a ton of merge conflicts here - can you rebase?

**Comment by deivishome** [comment]
Dear snipe, i am bit newby using github, how do i have to do a rebase? Thnkx for your help.

**Comment by deivishome** [comment]
OK rebased, can you check please? Thx.

**Comment by snipe** [comment]
I'm still seeing merge conflicts here - they need to resolved before I can merge.

**Comment by deivishome** [comment]
OK corrected

**Comment by snipe** [comment]
I still see 50 files changed tho? That doesn't seem right - and a lot of them are my changes. The PR should really be just your changes.

**Comment by deivishome** [comment]
yes its strange 22 css files and i only changed 4 files, i don't know if this can be reverted and discard the PR and i will start over

**Comment by deivishome** [comment]
ok snipe can you check it again please? sorry for the inconvinience

File: app/Http/Controllers/Api/LicensesController.php [hunk 1]
```diff
@@ -12,6 +12,10 @@
 use App\Models\LicenseSeat;
 use Illuminate\Http\Request;
 use Illuminate\Support\Facades\DB;
+use App\Http\Controllers\Api\User;
```
Comment (snipe) [inline app/Http/Controllers/Api/LicensesController.php:15]: I'm not sure why we'd be using a controller here?

File: app/Http/Controllers/Api/LicensesController.php [hunk 2]
```diff
@@ -304,6 +308,122 @@ public function selectlist(Request $request)
...
+           }
+            $this->authorize('checkout', License::class);
+           $licenseSeat = $this->findLicenseSeatToCheckout($license, $seatId);
+           $licenseSeat->user_id = $request->filled('checkout_to');
+
+           $checkoutMethod = 'checkoutTo'.ucwords(request('checkout_to_type'));
+           if ($this->$checkoutMethod($request, $licenseSeat)) {
+                return response()->json(Helper::formatStandardApiResponse('success', null, trans('admin/licenses/message.checkout.success')));
+           }
+           return response()->json(Helper::formatStandardApiResponse('error', null, trans('Something went wrong handling this checkout.')));
```
Comment (snipe) [inline app/Http/Controllers/Api/LicensesController.php:289]: I'm not sure this `trans()` is correct. It's looking for a matching key in the `resources/lang/en` directory, not an actual string

File: app/Models/License.php [hunk 3]
```diff
@@ -678,4 +678,21 @@ public function scopeOrderCompany($query, $order)
...
             ->orderBy('companies.name', $order);
     }
+    /**
+     * Determines if a license is available for checkout if there are enough free seats using API REST
+     *
+     * @author [D.Valin] [<david.valin@deivishome.com>]
+     * @since [v3.0]
+     * @return boolean
+     */
+    public function availableForCheckout()
```
Comment (snipe) [inline app/Models/License.php:688]: I believe we already have a method that does this. See `freeSeat()`

Comment (snipe) [reply app/Models/License.php:688]: (Also, please don't use capital letters for variable names.)

**Review by snipe — CHANGES_REQUESTED** [review]
Sorry this took a bit to get through. 

I have some questions here, but also I think this PR duplicates the functionality in the LicenseSeats API controller, in the `update()` method, so I'm not even sure it needs to be made?

File: app/Http/Controllers/Api/LicensesController.php (hunk 2 above)
Comment (deivishome) [reply app/Http/Controllers/Api/LicensesController.php:289]: Dear Snipe i set trans based on other api responses so if there is a translation of the language it should return the translated chain, if not, i will return the eng language chain, thats what i though.

Comment (snipe) [reply app/Http/Controllers/Api/LicensesController.php:289]: That's not how that works. In later versions of Laravel, you can do that using `__('this is my string')`, but we don't use that method currently.

File: app/Models/License.php (hunk 3 above)
Comment (deivishome) [reply app/Models/License.php:688]: Dear Snipe,
I used the remaincount as it returns the global value betwen available and checkedout, and to go faster i the remaicount > 0 then there are available, freeseats returns a list of information, and for the API in where you are just requesting to checkout a licenes, given the assgined_to_type (user, asset) and the value of the user or asset, and optionally note then i thought iwould be faster, if the funtion returns true then i check out to the seat/user id and type and Snipe associates it, if there is not available, return false then i would not allow to checkout, just having a true or false, not a list. But as you said then based on the checkout function from web interface and did not used, i should had removed. I did not used on api/license controller.

File: app/Http/Controllers/Api/LicensesController.php (hunk 1 above)
Comment (deivishome) [reply app/Http/Controllers/Api/LicensesController.php:15]: Dear Snipe,
on the the licensescontroller, as it uses user_id to refer to wich user the license is going to be checked on checkout funtion it complained when trying that this App\Http\Controllers\Api\User controller was not present, the log on laravel.log after setting the line the error dissapeared and the checkout worked.

**Comment by deivishome** [comment]
> 
> 
> Sorry this took a bit to get through.
> 
> I have some questions here, but also I think this PR duplicates the functionality in the LicenseSeats API controller, in the `update()` method, so I'm not even sure it needs to be made?

Dear Snipe,
the API update of the license (using PATCH method) does not have the checkout at least in the readme.io https://snipe-it.readme.io/reference#licensesid-2 , there you can update details of the license but not check out to a new user or asset or even check in the license to recover the seat.

<blockquote><div>Snipe-IT Documentation</div><div><strong><a href="https://snipe-it.readme.io/reference#licensesid-2">Overview</a></strong></div><div>Snipe-IT is a free, open source IT asset management system. Features include management of assets, users, licenses, accessories, consumables and components, as well as two-factor authentication, LDAP/AD syncing, and asset acceptance confirmation.</div></blockquote>

**Comment by laxlie** [comment]
@deivishome Do you know if documentation exists for this anywhere? I have been unable to find, but likely I missed the obvious. 
Side note: @snipe your project is amazing. This is my first job where I get to really focus on computer things, and its been awesome to work with your software!

**Comment by deivishome** [comment]
> 
> 
> @deivishome Do you know if documentation exists for this anywhere? I have been unable to find, but likely I missed the obvious.
> Side note: @snipe your project is amazing. This is my first job where I get to really focus on computer things, and its been awesome to work with your software!

Dear laxlie,
there is no documentation at web page, the functionality through the API is easy to understand once you know what you need, as it follows web gui requirements to checkout/checkin a license to user or asset, let me explain:

To checkout to user the url from API it should be like this:
http(S)://SERVERNAME/api/v1/licenses/4/checkout?checkout_to_type=user&assigned_to=3
where 4 is the license ID (in the example)
checkout? its the action you are going to do
checkout_to_type=user is for assign the license to user
assigned_to where 3 is the user id (in the example)

To checkout to asset the url from API it should be like this:
http(S)://SERVERNAME/api/v1/licenses/4/checkout?checkout_to_type=asset&asset_id=7
where 4 is the license ID (in the example)
checkout? its the action you are going to do
checkout_to_type=asset is for assign the license to user
asset_id where 7 is the asset id (in the example)

To checkin the license and get the seat free again (both user and asset):
http(S)://SERVERNAME/api/v1/licenses/4/checkin?seat_id=&back_to=
Leaving seat_id and back_to without values it will free 1 license seat.

I hope this helps.

Sorry for the late response as  had a family issue.

Best regards.

**Comment by beinhart** [comment]
Which version of SnipeIT containes the API license checkout functionality? With 5.2 I receive an "Method not allowed" error. See also https://github.com/snipe/snipe-it/issues/10309 ( Checkout a license to a user via API and fails #10309 )

**Comment by mahlquistj** [comment]
I'm also having the same issue as @beinhart

I'm running SnipeIT v5.3.8 - build 6619.

I've tried to POST `https://hw.hiper.dk/api/v1/licenses/<license_id>/checkout` with data:
`{
    "checkout_to_type": "user",
    "assigned_user": <user_id>
}`

I'm getting the following response:
`{
    "status": "error",
    "messages": "Method not allowed",
    "payload": null
}`

**Comment by deivishome** [comment]
> 

Dear Madser123,
in checkout the parameter is not assigned_user is assigned_to and then the id of the user. 
Try this please.

Best regards,

**Comment by mahlquistj** [comment]
@deivishome thank you for the reply.

I tried changing `"assigned_user": <user_id>` to `"assigned_to": <user_id>`

Still the same result unfortunately:
`{
    "status": "error",
    "messages": "Method not allowed",
    "payload": null
}`

I also tried with the user-id as a string, just for fun, but that doesn't work either, unfortunately :disappointed: 

I also tried the same request (`POST: /licenses/<license_id>/checkout`), but with the paramaters in the url-query instead: `https://<SERVERNAME>/api/v1/licenses/<licenses_id>/checkout?checkout_to_type=user&assigned_to=<user_id>`

Also still the same "Method not allowed" response.

**Comment by deivishome** [comment]
Hi Madser,
what I do is:

Dim REQ3 As New RestRequest
            Dim CLI3 As New RestClient("https://XXXXX.XXXXX.COM/api/v1/licenses/" & IDLIC & "/checkout")
            REQ3.Method = Method.POST
            REQ3.AddHeader("Authorization", "Bearer CODEHERE")
            REQ3.AddHeader("Accept", "application/json")
            REQ3.AddHeader("Content-type", "application/x-www-form-urlencoded")
            REQ3.AddHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:87.0) Gecko/20100101 Firefox/87.0")
            REQ3.AddParameter("checkout_to_type", "user")
            REQ3.AddParameter("assigned_to", IDUSER)
            Dim rep3 = CLI3.Execute(REQ3)
            Dim RESPU As String = ""
            RESPU = JsonHelper.GetJson(rep3.Content, "status", 1)

My Snipe-IT version is v5.1.4 build 5886 I did that development because the original API did not have that functionality but I don't know it in newer release it is coded or not, I had some troubles uploading the code to github and Snipe helped me.

**Comment by mahlquistj** [comment]
@snipe  - Can i ask wether or not this is being implemented, or should be, in SnipeIT version v5.3.8 - build 6619?

EDIT: I can open a feature request, if that's what you want me to do :smile:

**Pull request closed without merging**